### PR TITLE
Add support for factors API

### PR DESCRIPTION
**Pull request by bdemers**

### CODE REVIEW

**Comment by bdemers** [comment]
I just disabled codacy looks like it always reports an `x` because coverage is not enabled, (which we are using codecov for.  Enabling coverage for codacy looks easy but it requires yet another CI key, and codecov does not.... (I do like the health reports of 'codacy' though but they don't look like anything you cannot get from Findbug, PMD, or Checkstyle.  We could just tweak our Findbugs and PMD config (we are not currently using Checkstyle)

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule ".submodules/okta-sdk-test-server"]
```
Comment (bdemers) [inline .gitmodules]: Pulls in IT mock dependencies.  When this project is more stable, we can pull them in with webjars. Until then, this config, and the corresponding pom.xml, make using git submodules transparent (or as much as that is possible)

File: impl/pom.xml [hunk 2]
```diff
@@ -85,6 +85,13 @@
     </dependencies>
 
     <build>
+
+        <resources>
+            <resource>
+                <directory>target/generated-resources/swagger</directory>
```
Comment (bdemers) [inline impl/pom.xml]: Don't forget about the generated code!

File: impl/src/main/java/com/okta/sdk/impl/ds/DefaultResourceConverter.java [hunk 3]
```diff
@@ -74,16 +73,7 @@ private Object toMapValue(final AbstractResource resource, final String propName
...
-            //if the property is a reference, don't write the entire object - just the href will do:
-            //TODO need to change this to write the entire object because this code defeats the purpose of entity expansion
-            //     when this code gets called (returning the reference instead of the whole object that is returned from Okta)
-            if(Strings.hasText(resource.getResourceHref())) {
-                return this.referenceFactory.createReference(propName, (Map) value);
-            }
-            else{
-                return this.referenceFactory.createUnmaterializedReference(propName, (Map) value);
-            }
+            return this.referenceFactory.createUnmaterializedReference(propName, (Map) value);
```
Comment (bdemers) [inline impl/src/main/java/com/okta/sdk/impl/ds/DefaultResourceConverter.java]: This was a left over from the Stormpath SDK, with the Okta data model, everything is included directly in the model

File: impl/src/main/java/com/okta/sdk/impl/ds/DefaultResourceFactory.java [hunk 4]
```diff
@@ -47,10 +47,21 @@ public DefaultResourceFactory(InternalDataStore dataStore) {
             throw new NullPointerException("Resource class cannot be null.");
         }
 
-        Class<T> implClass = getImplementationClass(clazz);
+        Class<T> classToResolve = clazz;
 
         Object[] ctorArgs = createConstructorArgs(constructorArgs);
 
+        if (ctorArgs.length >= 2 && ctorArgs[1] instanceof Map) {
```
Comment (bdemers) [inline impl/src/main/java/com/okta/sdk/impl/ds/DefaultResourceFactory.java]: Check to see if there are multiple implementations.
With the Factors API, there are multiple types of factors, and the data map must be inspected before figuring out the correct interface

File: impl/src/main/java/com/okta/sdk/impl/security/DefaultSaltGenerator.java [hunk 5]
```diff
@@ -40,6 +40,6 @@ public String generate() {
         SecureRandom random = new SecureRandom();
         byte[] byteArr = new byte[bytesSize];
         random.nextBytes(byteArr);
-        return Base64.encodeBase64URLSafeString(byteArr);
+        return Base64.getUrlEncoder().encodeToString(byteArr);
```
Comment (bdemers) [inline impl/src/main/java/com/okta/sdk/impl/security/DefaultSaltGenerator.java]: Remove dead code below (Java8 has native support for this, finally)

File: impl/src/test/groovy/com/okta/sdk/impl/resource/GeneratedResourceTest.groovy [hunk 6]
```diff
@@ -48,12 +50,7 @@ class GeneratedResourceTest {
 
         InternalDataStore dataStore = createMockDataStore()
 
-        Arrays.stream(getClasses("com.okta.sdk.impl.resource"))
-            .filter { clazz -> clazz.superclass != null}
-            .filter { clazz -> clazz.superclass == AbstractInstanceResource}
-            .filter { clazz -> !clazz.name.contains("Test")}
-            .filter { clazz -> !clazz.name.contains("Abstract")}
-            .forEach { clazz ->
+        clazzStream().forEach { clazz ->
```
Comment (bdemers) [inline impl/src/test/groovy/com/okta/sdk/impl/resource/GeneratedResourceTest.groovy]: DRY this up

File: integration-tests/pom.xml [hunk 7]
```diff
@@ -63,39 +63,37 @@
...
-                </excludes>
-            </testResource>
-        </testResources>
-    </build>
-
     <profiles>
         <profile>
-            <!-- This profile configures the Node.js based test-server -->
-            <id>sdk-test-server</id>
+            <id>mock-server</id>
```
Comment (bdemers) [inline integration-tests/pom.xml]: SDK testing mock server has been replaced with a java implementation (that uses wiremock)

File: integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/GroupRuleIT.groovy [hunk 8]
```diff
@@ -94,7 +94,7 @@ class GroupRuleIT extends ITSupport {
         rule.activate()
 
         // Triggering the updated rule will remove the user from group
-        assertUserInGroup(user, group, 20, 1000, false)
+        assertUserInGroup(user, group, 120, 1000, false)
```
Comment (bdemers) [inline integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/GroupRuleIT.groovy]: flaky test, depending on server timing, using mock responses will help with this in the near future

File: integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/UserIT.groovy [hunk 9]
```diff
@@ -130,6 +132,10 @@ class UserIT extends ITSupport {
             .setOldPassword(client.instantiate(PasswordCredential).setValue('Abcd1234'))
             .setNewPassword(client.instantiate(PasswordCredential).setValue('1234Abcd')))
         assertThat credentials.provider.type, equalTo(AuthenticationProviderType.OKTA)
+
+        // 3. make the test recording happy, and call a get on the user
+        // TODO: fix har file
+        client.getUser(user.id)
```
Comment (bdemers) [inline integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/UserIT.groovy]: @vijetmahabaleshwar-okta, i'm not sure why these are needed, is there some additional validation that some of the other SDKs are doing

File: integration-tests/src/test/groovy/com/okta/sdk/tests/it/util/ClientProvider.groovy [hunk 10]
```diff
@@ -64,7 +64,7 @@ trait ClientProvider implements IHookable {
     private Client buildClient(String scenarioId = null) {
 
         String testServerBaseUrl = System.getProperty(TestServer.TEST_SERVER_BASE_URL)
-        if (isRunningWithTestServer()) {
+        if (isRunningWithTestServer() && scenarioId != null) {
```
Comment (bdemers) [inline integration-tests/src/test/groovy/com/okta/sdk/tests/it/util/ClientProvider.groovy]: Don't use the mock-server if the `scenarioId` is not set

File: pom.xml [hunk 11]
```diff
@@ -236,11 +237,74 @@
...
+        </plugins>
     </build>
+    <profiles>
+        <profile>
+            <id>ci</id>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>com.github.siom79.japicmp</groupId>
+                        <artifactId>japicmp-maven-plugin</artifactId>
```
Comment (bdemers) [inline pom.xml]: Runs a report (manually) to see the delta between versions, once we hit 1.0 we can enable this in CI

File: src/findbugs/findbugs-exclude.xml [hunk 12]
```diff
@@ -64,4 +64,10 @@
     <!-- Quickstart is formatted as an example, so there is dead code and hardcoded passwords -->
     <Match><Class name="quickstart.Quickstart"/></Match>
 
+    <!-- path destination is passed in as a build parameter -->
+    <Match>
+        <Package name="com.okta.swagger.codegen"></Package>
+        <Bug pattern="PATH_TRAVERSAL_IN" />
```
Comment (bdemers) [inline src/findbugs/findbugs-exclude.xml]: Used at build time, NOT runtime

File: impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.resource
+
+import org.testng.annotations.Test
+import static org.hamcrest.Matchers.*
```
Comment (jmelberg-okta) [inline impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy]: Does this need the wildcard import?

File: impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            Map<String, Property> getPropertyDescriptors() {
+                return Collections.emptyMap()
+            }
+        }
+
+        resource.setResourceHref(href)
+        def actualMap = resource.getParamsFromHref(templateUrl)
+        assertThat actualMap, equalTo(expectedValues)
+    }
+}
```
Comment (jmelberg-okta) [inline impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy]: nit: newlines

File: impl/src/main/java/com/okta/sdk/impl/resource/AbstractResource.java [hunk 15]
```diff
@@ -664,4 +670,46 @@ public void clear() {
...
+        Assert.hasText(templateUrl, "Template URL must not be empty.");
+        Map<String, String> resultMap = new HashMap<>();
+
+        try {
+            String hrefPath = new URI(href).getPath();
+
+            Iterator<String> templatePartsIter = Arrays.asList(templateUrl.split("/")).iterator();
+            Iterator<String> hrefPartsIter = Arrays.asList(hrefPath.split("/")).iterator();
+
+            while ( templatePartsIter.hasNext() && hrefPartsIter.hasNext()) {
```
Comment (jmelberg-okta) [inline impl/src/main/java/com/okta/sdk/impl/resource/AbstractResource.java]: nit: extra space

Comment (bdemers) [reply impl/src/main/java/com/okta/sdk/impl/resource/AbstractResource.java]: 👍

File: impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy (hunk 13 above)
Comment (bdemers) [reply impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy]: No love for the test wildcards from anybody :P I just keep forgetting to clean them up after writing them (`org.hamcrest.Matchers`) just contains a bunch of static methods

File: impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy (hunk 14 above)
Comment (bdemers) [reply impl/src/test/groovy/com/okta/sdk/impl/resource/AbstractResourceTest.groovy]: some ask for them to be removed, others added :P

File: integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/UserIT.groovy (hunk 9 above)
Comment (vijetmahabaleshwar-okta) [reply integration-tests/src/test/groovy/com/okta/sdk/tests/it/spec/UserIT.groovy]: @bdemers - I think after changing password, we're validating that 'last password change time' has increased by calling getUser again.

**Pull request merged**